File: cpp/src/internal/Token.cpp
#include "internal/Token.h"
#include <map>
#include <iostream>

namespace {
  inline int compare(int x, int y) {
    return (x < y) ? -1 : ((x == y) ? 0 : 1);
  }

  struct StringToken : Token {
    explicit StringToken(const std::string& str) : str(str) {
    }

  private:
    int safeCompareTo(const Token& token) const override {
      auto& other = dynamic_cast<const StringToken&>(token);
      return compare(str.compare(other.str), 0);
    }

  private:
    std::string str;
  };

  struct NumericToken : Token {
    NumericToken(int length, int value) : length(length), value(value) {
    }

  private:
    int safeCompareTo(const Token& token) const override {
      auto& other = dynamic_cast<const NumericToken&>(token);
      auto result = compare(value, other.value);
      return result == 0 ? compare(length, other.length) : result;
    }

  private:
    int length;
    int value;
  };

  std::map<std::string, int> priorities = {
      { typeid(NumericToken).name(), 1 },
      { typeid(StringToken).name(),  2 },
  };
}

int Token::compareTo(const Token& other) const {
  auto priority = compare(
      priorities[typeid(*this).name()],
      priorities[typeid(other).name()]);
  return priority == 0 ? safeCompareTo(other) : priority;
}

Token* Token::newStringToken(const std::string& str) {
  return new StringToken(str);
}

Token* Token::newNumericToken(int length, int value) {
  return new NumericToken(length, value);
}

File: java/src/main/java/coach2018/Token.java
package coach2018;

import java.util.HashMap;
import java.util.Map;

import static java.lang.Integer.compare;

abstract class Token implements Comparable<Token> {
    private static Map<Class<?>, Integer> priorities = new HashMap<Class<?>, Integer>() {{
        put(NumericToken.class, 1);
        put(StringToken.class,  2);
    }};

    @Override
    public int compareTo(Token other) {
        int priority = compare(
                priorities.get(getClass()),
                priorities.get(other.getClass()));
        return priority == 0 ? safeCompareTo(other) : priority;
    }

    abstract protected int safeCompareTo(Token other);
}

class StringToken extends Token {
    private String str;

    StringToken(String str) {
        this.str = str;
    }

    @Override
    protected int safeCompareTo(Token token) {
        StringToken other = (StringToken)token;
        return compare(str.compareTo(other.str), 0);
    }
}

class NumericToken extends Token {
    private int length;
    private int value;

    public NumericToken(int length, int value) {
        this.length = length;
        this.value = value;
    }

    @Override
    protected int safeCompareTo(Token token) {
        NumericToken other = (NumericToken)token;
        int result = compare(value, other.value);
        return result == 0 ? compare(other.length, length) : result;
    }
}
